keras_trainer/trainer.py

```diff
@@ -85,44 +90,51 @@ def run(self):
-                'pooling': self.pooling
-            }
-            model_kwargs.update(self.model_kwargs)
-            model = self.model_spec.klass(**model_kwargs)
-
-            # Expand the base model to match the spec target_size.
-            model_kwargs.update({
-                'input_shape': self.model_spec.target_size,
-                'weights': None
-            })
-            expanded_model = self.model_spec.klass(**model_kwargs)
-            for i in range(1, len(expanded_model.layers)):
-                expanded_model.layers[i].set_weights(model.layers[i].get_weights())
-            model = expanded_model
+        if self.checkpoint_path is not None:
+            model = load_model(self.checkpoint_path)
         else:
-            model = self.model_spec.klass(
-                input_shape=self.model_spec.target_size,
-                weights=self.weights,
-                include_top=False,
-                pooling=self.pooling
-            )
-
-        # Include the given top layers.
-        if self.top_layers is None:
-            layer = Dense(self.num_classes, name='dense')(model.output)
-            self.top_layers = Activation('softmax', name='act_softmax')(layer)
-        model = Model(model.input, self.top_layers)
+            if self.model_spec.klass == MobileNet:
+                # Initialize the base with valid target_size.
+                model_kwargs = {
+                    'input_shape': (224, 224, 3),
+                    'weights': self.weights,
+                    'include_top': False,
+                    'pooling': self.pooling
+                }
+                model_kwargs.update(self.model_kwargs)
+                model = self.model_spec.klass(**model_kwargs)
+
+                # Expand the base model to match the spec target_size.
+                model_kwargs.update({
+                    'input_shape': self.model_spec.target_size,
+                    'weights': None
+                })
+                expanded_model = self.model_spec.klass(**model_kwargs)
+                for i in range(1, len(expanded_model.layers)):
+                    expanded_model.layers[i].set_weights(model.layers[i].get_weights())
+                model = expanded_model
+            else:
+                model = self.model_spec.klass(
+                    input_shape=self.model_spec.target_size,
+                    weights=self.weights,
+                    include_top=False,
+                    pooling=self.pooling
+                )
+
+            # Include the given top layers.
+            if self.top_layers is None:
+                layer = Dense(self.num_classes, name='dense')(model.output)
+                self.top_layers = Activation('softmax', name='act_softmax')(layer)
+            model = Model(model.input, self.top_layers)
 
         # Print the model summary.
         if self.verbose:
             model.summary()
 
+        # GPU multiprocessing
+        if self.num_gpu > 1:
```

Comment:
Can `num_gpu` be detected with something like this as a default?

```python
from tensorflow.python.client import device_lib

def get_available_gpus():
    local_device_protos = device_lib.list_local_devices()
    return [x.name for x in local_device_protos if x.device_type == 'GPU']
```

Reply:
But this would use all the available GPUs while in some cases we would like to distribute different jobs between them, what do you think?
Maybe we can use it to set the maximum.

Reply:
In most cases I think we want to make use of all available GPUs by default right? So how about if no (or `None`) `num_gpus` is passed in then it detects and uses all available, but if a non-None `num_gpus` is passed in as an override then it uses that?

Reply:
I was thinking in the case we have a shared large cluster with a big number of GPUs, so we may need to distribute them between us. I like your last approach, I'll implement that option.